azure/services/privatedns/spec.go

```diff
@@ -0,0 +1,188 @@
+		})),
+	}, nil
+}
+
+// ResourceName returns the name of the virtual network link.
+func (l LinkSpec) ResourceName() string {
+	return l.Name
+}
+
+// OwnerResourceName returns the zone name of the virtual network link.
+func (l LinkSpec) OwnerResourceName() string {
+	return l.ZoneName
+}
+
+// ResourceGroupName returns the name of the resource group of the virtual network link.
+func (l LinkSpec) ResourceGroupName() string {
+	return l.ResourceGroup
+}
+
+// Parameters returns the parameters for the virtual network link.
+func (l LinkSpec) Parameters(existing interface{}) (params interface{}, err error) {
+	_, log, done := tele.StartSpanWithLogger(context.TODO(), "privatedns.LinkSpec.Parameters")
+	defer done()
+
+	if existing != nil {
+		link, ok := existing.(privatedns.VirtualNetworkLink)
+		if !ok {
+			return nil, errors.Errorf("%T is not a privatedns.VirtualNetworkLink", existing)
+		}
+		if !isManaged(link.Tags, l.ClusterName) {
+			log.V(2).Info("Skipping vnet link reconciliation for unmanaged vnet link", "vnet link",
+				l.Name, "private dns zone", l.ZoneName)
+			return nil, nil
+		}
+	}
+
+	return privatedns.VirtualNetworkLink{
+		VirtualNetworkLinkProperties: &privatedns.VirtualNetworkLinkProperties{
+			VirtualNetwork: &privatedns.SubResource{
+				ID: to.StringPtr(azure.VNetID(l.SubscriptionID, l.VNetResourceGroup, l.VNetName)),
+			},
+			RegistrationEnabled: to.BoolPtr(false),
+		},
+		Location: to.StringPtr(azure.Global),
+		Tags: converters.TagsToMap(infrav1.Build(infrav1.BuildParams{
+			ClusterName: l.ClusterName,
+			Lifecycle:   infrav1.ResourceLifecycleOwned,
+			Additional:  l.AdditionalTags,
+		})),
+	}, nil
+}
+
+// ResourceName returns the name of the virtual network link.
+func (r RecordSpec) ResourceName() string {
+	return r.Record.Hostname
+}
+
+// OwnerResourceName returns the zone name of the virtual network link.
+func (r RecordSpec) OwnerResourceName() string {
+	return r.ZoneName
+}
+
+// ResourceGroupName returns the name of the resource group of the virtual network link.
+func (r RecordSpec) ResourceGroupName() string {
+	return r.ResourceGroup
+}
+
+// Parameters returns the parameters for the virtual network link.
+func (r RecordSpec) Parameters(existing interface{}) (params interface{}, err error) {
+	set := privatedns.RecordSet{
+		RecordSetProperties: &privatedns.RecordSetProperties{
+			TTL: to.Int64Ptr(300),
+		},
+	}
+	recordType := converters.GetRecordType(r.Record.IP)
+	if recordType == privatedns.A {
+		set.RecordSetProperties.ARecords = &[]privatedns.ARecord{{
+			Ipv4Address: &r.Record.IP,
+		}}
+	} else if recordType == privatedns.AAAA {
+		set.RecordSetProperties.AaaaRecords = &[]privatedns.AaaaRecord{{
+			Ipv6Address: &r.Record.IP,
+		}}
+	}
+
+	return set, nil
+}
+
+// isManaged returns true if tagsMap contains an owner tag for the cluster.
+func isManaged(tagsMap map[string]*string, clusterName string) bool {
+	tags := converters.MapToTags(tagsMap)
+	return tags.HasOwned(clusterName)
+}
```

Comment:
nit: 
This could move to a different package probably -- which can help reuse.

Reply:
Sure. May be we can put a comment there by putting a TODO. This can help track in future if we want to reuse and refactor.